core/src/main/java/org/elasticsearch/ingest/PipelineStore.java

```diff
@@ -154,18 +115,58 @@ public void close() throws IOException {
         IOUtils.close(closeables);
     }
 
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        IngestMetadata ingestMetadata = event.state().getMetaData().custom(IngestMetadata.TYPE);
+        if (ingestMetadata == null) {
+            return;
+        }
+
+        Map<String, Pipeline> pipelines = new HashMap<>();
+        for (Map.Entry<String, Map<String, Object>> entry : ingestMetadata.getPipelines().entrySet()) {
+            String id = entry.getKey();
+            try {
+                @SuppressWarnings("unchecked")
+                Map<String, Object> pipelineConfigCopy = (Map<String, Object>) deepCopy(entry.getValue());
+                pipelines.put(id, constructPipeline(id, pipelineConfigCopy));
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }
+        this.pipelines = Collections.unmodifiableMap(pipelines);
+    }
+
     /**
      * Deletes the pipeline specified by id in the request.
      */
-    public void delete(DeletePipelineRequest request, ActionListener<DeleteResponse> listener) {
-        ensureReady();
+    public void delete(DeletePipelineRequest request, ActionListener<DeletePipelineResponse> listener) {
+        clusterService.submitStateUpdateTask("delete-pipeline-" + request.id(), new AckedClusterStateUpdateTask<DeletePipelineResponse>(request, listener) {
+
+            @Override
+            protected DeletePipelineResponse newResponse(boolean acknowledged) {
+                return new DeletePipelineResponse(acknowledged);
+            }
 
-        DeleteRequest deleteRequest = new DeleteRequest(request);
-        deleteRequest.index(PipelineStore.INDEX);
-        deleteRequest.type(PipelineStore.TYPE);
-        deleteRequest.id(request.id());
-        deleteRequest.refresh(true);
-        client.delete(deleteRequest, handleWriteResponseAndReloadPipelines(listener));
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                IngestMetadata currentIngestMetadata = currentState.metaData().custom(IngestMetadata.TYPE);
```

Comment:
would be nice to tests this without a ClusterService, can we make all the execute bodies pkg private methods we can call and have dedicated unittests and assert the content etc?